ot/da.py

```diff
@@ -843,10 +927,896 @@ def predict(self, x):
+                                     'with `estimator.get_params().keys()`.' %
+                                     (key, self.__class__.__name__))
+                setattr(self, key, value)
+        return self
+
+    def __repr__(self):
+        from sklearn.base import _pprint
+        class_name = self.__class__.__name__
+        return '%s(%s)' % (class_name, _pprint(self.get_params(deep=False),
+                                               offset=len(class_name),),)
+
+    # __getstate__ and __setstate__ are omitted because they only contain
+    # conditionals that are not satisfied by our objects (e.g.,
+    # ``if type(self).__module__.startswith('sklearn.')``.
+
+
+def distribution_estimation_uniform(X):
+    """estimates a uniform distribution from an array of samples X
+
+    Parameters
+    ----------
+    X : array-like, shape (n_samples, n_features)
+        The array of samples
+
+    Returns
+    -------
+    mu : array-like, shape (n_samples,)
+        The uniform distribution estimated from X
+    """
+
+    return unif(X.shape[0])
+
+
+class BaseTransport(BaseEstimator):
+
+    def fit(self, Xs=None, ys=None, Xt=None, yt=None):
+        """Build a coupling matrix from source and target sets of samples
+        (Xs, ys) and (Xt, yt)
+
+        Parameters
+        ----------
+        Xs : array-like, shape (n_source_samples, n_features)
+            The training input samples.
+        ys : array-like, shape (n_source_samples,)
+            The class labels
+        Xt : array-like, shape (n_target_samples, n_features)
+            The training input samples.
+        yt : array-like, shape (n_labeled_target_samples,)
+            The class labels
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+
+        # pairwise distance
+        self.cost_ = dist(Xs, Xt, metric=self.metric)
+
+        if (ys is not None) and (yt is not None):
+
+            if self.limit_max != np.infty:
+                self.limit_max = self.limit_max * np.max(self.cost_)
+
+            # assumes labeled source samples occupy the first rows
+            # and labeled target samples occupy the first columns
+            classes = np.unique(ys)
+            for c in classes:
+                idx_s = np.where((ys != c) & (ys != -1))
+                idx_t = np.where(yt == c)
+
+                # all the coefficients corresponding to a source sample
+                # and a target sample :
+                # with different labels get a infinite
+                for j in idx_t[0]:
+                    self.cost_[idx_s[0], j] = self.limit_max
+
+        # distribution estimation
+        self.mu_s = self.distribution_estimation(Xs)
+        self.mu_t = self.distribution_estimation(Xt)
+
+        # store arrays of samples
+        self.Xs = Xs
+        self.Xt = Xt
+
+        return self
+
+    def fit_transform(self, Xs=None, ys=None, Xt=None, yt=None):
```

Comment:
SAme here with the necessary parameters, maybe we should coce a test function with an error message that lists the necessary parameters i'm thinking as 

```python

def check_params(**kwargs):
    OK=True
    for param in kwargs:
        if kwargs[kwargs] is None:
            OK=False

    if not OK:
        #Print the list of Necessary params in kwargs

    return OK


# the function can be called inside the classes as
if check_params(Xs=Xs,Xt=Xt):
    do the stuff

```
What is nice is that we can give different keywords params depending on the class (semi supervised  or else).

This function or at leats a clener version should be in utils.py

Reply:
ok

Reply:
Sorry but this is not what you did in the last commit, you just added a condition, oi think the implementation above would be more future proof since it handles the warning/error message automatically, no need for else.